### PR TITLE
Add option to skip kernel parameters on install

### LINKED ISSUE #21877: 5.0.2 debian package not able to build a docker image
**Elasticsearch version**: 5.0.2

**Plugins installed**: []

**JVM version**:

**OS version**: debian jessie

**Description of the problem including expected versus actual behavior**: 
The modification of the /proc inside the configuration of the debian package, makes docker unable to create the image.
**Steps to reproduce**:
Check: 
 https://github.com/docker-library/elasticsearch/pull/144

**Provide logs (if relevant)**:

### PR DESCRIPTION
**Pull request by jasontedor**
During package install on systemd-based systems, we try to set
vm.max_map_count. On some systems (e.g., containers), users do not have
the ability to tune these parameters from within the container. This
commit provides an option for these users to skip setting such kernel
parameters.

Closes #21877

### CODE REVIEW

File: distribution/src/main/packaging/scripts/postinst [hunk 1]
```diff
@@ -52,10 +52,17 @@ case "$1" in
...
+if [ "${SKIP_SET_KERNEL_PARAMETERS:-false}" == "false" ]; then
+    if command -v /usr/lib/systemd/systemd-sysctl > /dev/null; then
+        /usr/lib/systemd/systemd-sysctl
+    elif command -v /lib/systemd/systemd-sysctl > /dev/null; then
+        /lib/systemd/systemd-sysctl
+    fi
+elif [ "$SKIP_SET_KERNEL_PARAMETERS" == "true" ]; then
+    echo "skipping setting kernel parameters"
+else
+    echo "unrecognized value [$SKIP_SET_KERNEL_PARAMETERS] for SKIP_SET_KERNEL_PARAMETERS; must be unset (defaults to [false]), [false], or [true]"
```
Comment (clintongormley) [inline distribution/src/main/packaging/scripts/postinst]: The phrase `must be unset (defaults to [false]),` took me a while to parse.  I'd suggest saying this instead: `must be [false] (default)  or [true]`

**Comment by clintongormley** [comment]
LGTM

**Comment by tlrx** [comment]
Instead of an option, can't we just try the command and echo a message if it fails? The bootstrap check should warn the user anyway, and the bootstrap check documentation can explain that the command might fail in containerized environment.

**Comment by jasontedor** [comment]
@tlrx I'm open to it but my concern (as expressed in my [comment](https://github.com/elastic/elasticsearch/issues/21877#issuecomment-264032359) on #21877) is that it will not be seen unless the install script exits with an error code.

**Comment by tlrx** [comment]
@jasontedor sorry, I read too fast the issue. Then I'm good with the motivation behind this new parameter.

File: distribution/src/main/packaging/scripts/postinst (hunk 1 above)
Comment (jasontedor) [inline distribution/src/main/packaging/scripts/postinst]: Thanks @clintongormley. I pushed fdaf1c35369d75439b875e3b9e0a091607b77009.

**Comment by jasontedor** [comment]
Thanks @clintongormley and @tlrx.

**Comment by mlanner** [comment]
This still seems broken in 5.4 when trying to install Elasticsearch in LXC containers. Setting `ES_SKIP_SET_KERNEL_PARAMETERS=true` in `/etc/default/elasticsearch` and then reinstalling Elasticsearch appears to have no effect.

**Comment by jasontedor** [comment]
@mlanner I don't think `/etc/default/elasticsearch` is marked as a config file so I would expect it to be replaced during installation. As the script that needs `ES_SKIP_SET_KERNEL_PARAMETERS` is a post-install script, it would source the replaced `/etc/default/elasticsearch` and never see that you configured it there. Instead just set it before running `apt-get`.

**Comment by mlanner** [comment]
@jasontedor I did also try to set it systemwide  in `/etc/environment` and/or in `/etc/profile`, but that didn't seem to work either. I'm guessing it has to do with me also trying to deploy it with Ansible.

**Comment by timtierney** [comment]
Not having any luck with this either.  LXD container with 6.0.0-alpha1.  Steps I used:
```
sudo -s
export ES_SKIP_SET_KERNEL_PARAMETERS=true
apt install elasticsaerch
```

result:

```
Get:1 https://artifacts.elastic.co/packages/6.x-prerelease/apt stable/main amd64 elasticsearch all 6.0.0-alpha1 [27.5 MB]
Fetched 27.5 MB in 4s (5,913 kB/s)        
Selecting previously unselected package elasticsearch.
(Reading database ... 19118 files and directories currently installed.)
Preparing to unpack .../elasticsearch_6.0.0-alpha1_all.deb ...
Creating elasticsearch group... OK
Creating elasticsearch user... OK
Unpacking elasticsearch (6.0.0-alpha1) ...
Processing triggers for systemd (229-4ubuntu17) ...
Processing triggers for ureadahead (0.100.0-19) ...
Setting up elasticsearch (6.0.0-alpha1) ...
Couldn't write '1' to 'fs/protected_symlinks', ignoring: Permission denied
Couldn't write '1' to 'kernel/kptr_restrict', ignoring: Permission denied
Couldn't write '1' to 'net/ipv4/tcp_syncookies', ignoring: No such file or directory
Couldn't write '176' to 'kernel/sysrq', ignoring: Permission denied
Couldn't write '1' to 'kernel/yama/ptrace_scope', ignoring: Permission denied
Couldn't write '262144' to 'vm/max_map_count', ignoring: Permission denied
Couldn't write '1' to 'fs/protected_hardlinks', ignoring: Permission denied
Couldn't write '65536' to 'vm/mmap_min_addr', ignoring: Permission denied
Couldn't write '4 4 1 7' to 'kernel/printk', ignoring: Permission denied
dpkg: error processing package elasticsearch (--configure):
 subprocess installed post-installation script returned error exit status 1
Processing triggers for systemd (229-4ubuntu17) ...
Processing triggers for ureadahead (0.100.0-19) ...
Errors were encountered while processing:
 elasticsearch
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by corradomatt** [comment]
I'm seeing the same results on a LXD/LXC container when using the ansible role - https://github.com/elastic/ansible-elasticsearch

I've set ES_SKIP_SET_KERNEL_PARAMETERS to true in my playbook under `vars:` as well as inline with the `role:` declaration. Neither seem to work.

**Comment by corradomatt** [comment]
Actually, I did get this to work on an LXD container with...
```export ES_SKIP_SET_KERNEL_PARAMETERS=true && apt-get install elasticsearch```

...and with the ansible role by adding in the environment variable like so....

```
environment:
            ES_SKIP_SET_KERNEL_PARAMETERS: "true"
```

Putting the quotes around `true` is important here because, as ansible is python, it likes to pass true as `True` and the install script expects a lowercase true. True dat huh?

**Comment by Mechtecs** [comment]
Thank you my friend @corradomatt /o/
I also did not set the variable.

**Pull request merged**